lib/auth/machineid/machineidv1/spiffe_federation_syncer.go

```diff
@@ -0,0 +1,631 @@
+			return "bundle_source_changed"
+		}
+	}
+
+	return ""
+}
+
+func (s *SPIFFEFederationSyncer) syncTrustDomain(
+	ctx context.Context, name string,
+) (out *machineidv1.SPIFFEFederation, err error) {
+	ctx, span := tracer.Start(ctx, "SPIFFEFederationSyncer/syncTrustDomain")
+	defer func() {
+		tracing.EndSpan(span, err)
+	}()
+
+	current, err := s.cfg.Store.GetSPIFFEFederation(ctx, name)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to get SPIFFEFederation")
+	}
+
+	log := s.cfg.Logger.With(
+		"current_revision", current.GetMetadata().GetRevision(),
+		"trust_domain", current.GetMetadata().GetName(),
+	)
+
+	td, err := spiffeid.TrustDomainFromString(current.GetMetadata().GetName())
+	if err != nil {
+		return nil, trace.Wrap(err, "parsing metadata.name as trust domain name")
+	}
+
+	// Determine - should we sync...
+	syncReason := shouldSyncTrustDomain(ctx, log, s.cfg.Clock, current)
+	if syncReason == "" {
+		log.DebugContext(ctx, "Skipping sync as is not required")
+		return current, nil
+	}
+	log.InfoContext(ctx, "Sync starting", "reason", syncReason)
+
+	// Clone the persisted resource so we can compare to it.
+	out = proto.Clone(current).(*machineidv1.SPIFFEFederation)
+
+	// Refresh...
+	if out.Status == nil {
+		out.Status = &machineidv1.SPIFFEFederationStatus{}
+	}
+
+	var bundle *spiffebundle.Bundle
+	var nextSyncIn time.Duration
+	switch {
+	case current.Spec.BundleSource.HttpsWeb != nil:
+		url := current.Spec.BundleSource.HttpsWeb.BundleEndpointUrl
+		log.DebugContext(
+			ctx,
+			"Fetching bundle using https_web profile",
+			"url", url,
+		)
+		bundle, err = federation.FetchBundle(ctx, td, url, s.cfg.SPIFFEFetchOptions...)
+		if err != nil {
+			return nil, trace.Wrap(err, "fetching bundle using https_web profile")
+		}
+
+		// Calculate the duration before we should next sync, applying any
+		// sensible upper and lower bounds.
+		nextSyncIn = s.cfg.DefaultSyncInterval
+		if refreshHint, ok := bundle.RefreshHint(); ok {
+			if refreshHint < s.cfg.MinSyncInterval {
+				log.DebugContext(
+					ctx,
+					"Refresh hint is less than MinSyncInterval, using MinSyncInterval",
+					"refresh_hint", refreshHint,
+					"min_sync_interval", s.cfg.MinSyncInterval,
+				)
+				nextSyncIn = s.cfg.MinSyncInterval
+			} else if refreshHint > s.cfg.MaxSyncInterval {
+				log.DebugContext(
+					ctx,
+					"Refresh hint is greater than MaxSyncInterval, using MaxSyncInterval",
+					"refresh_hint", refreshHint,
+					"max_sync_interval", s.cfg.MaxSyncInterval,
+				)
+				nextSyncIn = s.cfg.MaxSyncInterval
+			} else {
+				nextSyncIn = refreshHint
+			}
+		}
+	case current.Spec.BundleSource.Static != nil:
+		log.DebugContext(
+			ctx, "Fetching bundle using spec.bundle_source.static.bundle",
+		)
+		bundle, err = spiffebundle.Parse(
+			td, []byte(current.Spec.BundleSource.Static.Bundle),
+		)
+		if err != nil {
+			return nil, trace.Wrap(
+				err, "parsing bundle from static profile",
+			)
+		}
```

Comment:
It seems that for `current.Spec.BundleSource.Static`  this function performs a NOP operation except for updating the timestamp. Is this intended? It might give the misleading impression that the static bundles are being updated synced.

Reply:
It's intended - the value within "spec.bundle_source.static.bundle" is effectively being copied into "status.current_bundle". Essentially, this accounts for a Patch/Update operation on "spec.bundle_source.static.bundle".